Core/src/ExponentialBackoff.php

```diff
@@ -75,10 +106,24 @@ public function execute(callable $function, array $arguments = [])
         $retryAttempt = 0;
         $exception = null;
 
+        // The $arguments are passed by reference
+        // thus are modifiable before the execution starts.
+        if ($this->onExecutionStartFunction) {
+            call_user_func_array($this->onExecutionStartFunction, [&$arguments]);
+        }
+
         while (true) {
             try {
                 return call_user_func_array($function, $arguments);
             } catch (\Exception $exception) {
+                if ($this->onRetryExceptionFunction) {
+                    // The $arguments are passed by reference so that the user has the ability to modify
+                    // some elements of the request on every retry(for example headers).
+                    call_user_func_array(
+                        $this->onRetryExceptionFunction,
+                        [$exception, $retryAttempt, &$arguments]
+                    );
```

Comment:
So, majorly the difference lies in the fact that the `retryFunction` is responsible to decide if the retry should take place or not.

While consider the `onRetryExceptionFunction` is a callback that will be called whenever an exception occurs inside this flow.

We could mix the 2 for sure, but mixing these 2 responsibilities doesn't feel right. Consider the resumable downloads:
When an exception occurs, we are supposed to tabulate the `Range headers` to be sent to the server, so that we can resume the download from that specific byte.

At the same time we also need to decide whether the error was resumable.

The former is only supposed to happen in case of downloads, while the latter happens for almost all requests in Storage with this PR.

Hope this provides some clarity.

Reply:
Resolving this :)